lib/spack/spack/detection/path.py

```diff
@@ -241,15 +241,19 @@ def by_executable(
         path_hints: list of paths to be searched. If None the list will be
             constructed based on the PATH environment variable.
     """
-    path_hints = spack.util.environment.get_path("PATH") if path_hints is None else path_hints
+    given_path_hints = path_hints is None
+    path_hints = (
+        spack.util.environment.get_path("PATH") if given_path_hints is None else path_hints
+    )
     exe_pattern_to_pkgs = collections.defaultdict(list)
     for pkg in packages_to_check:
         if hasattr(pkg, "executables") and hasattr(pkg, "platform_executables"):
             for exe in pkg.platform_executables():
                 exe_pattern_to_pkgs[exe].append(pkg)
-        # Add Windows specific, package related paths to the search paths
-        path_hints.extend(compute_windows_user_path_for_package(pkg))
-        path_hints.extend(compute_windows_program_path_for_package(pkg))
+        if given_path_hints:
```

Comment:
Is there any reason to add this conditional?

Reply:
Otherwise we search non specified paths when `path_hints` is given. By this point, `path_hints` is always truthy, but those extra paths are package dependent, so they need to be invoked at that point, but querying `path_hints` is meaningless. The conditional stores whether `path_hints` was provided and then prevents the extra search paths from being considered if so.
The prior behavior was incorrect.